### PR TITLE
runtime: move closures out of check_balance to refactor it slightly

### PR DESCRIPTION
**Pull request by mina86**
check_balance function is a pretty long one.  Attempt to shorten it
a little by moving closures out of it and converting them to
functions.  This does require a few more arguments to be passed to
those functions but I’d argue that’s beneficial since now it’s clear
what data each of the functions actually use without having to
analyse it carefully.

### CODE REVIEW

File: runtime/runtime/src/balance_checker.rs [hunk 1]
```diff
@@ -17,6 +17,99 @@ use near_primitives::version::ProtocolVersion;
...
+    accounts_ids.iter().try_fold(0u128, |accumulator, account_id| {
+        let (amount, locked) = match get_account(state, account_id)? {
+            None => return Ok(accumulator),
+            Some(account) => (account.amount(), account.locked()),
+        };
+        Ok(accumulator)
+            .and_then(|accumulator| safe_add_balance(accumulator, amount))
+            .and_then(|accumulator| safe_add_balance(accumulator, locked))
+            .map_err(|_| RuntimeError::UnexpectedIntegerOverflow)
+    })
```
Comment (matklad) [inline runtime/runtime/src/balance_checker.rs:93]: Let's maybe add 

```rust
pub fn safe_add_balances<const N: usize>(xs: [Balance; N]) -> Result<Balance, IntegerOverflowError>
```

?

File: runtime/runtime/src/balance_checker.rs [hunk 2]
```diff
@@ -17,6 +17,99 @@ use near_primitives::version::ProtocolVersion;
 use near_store::{get, get_account, get_postponed_receipt, TrieUpdate};
 use std::collections::HashSet;
 
+/// Returns delayed receipts with given range of indices.
+fn get_delayed_receipts(
+    state: &TrieUpdate,
+    from_index: u64,
+    to_index: u64,
+) -> Result<Vec<Receipt>, StorageError> {
```
Comment (matklad) [inline runtime/runtime/src/balance_checker.rs:25]: An alternative is to take an `indexes: Range<u64>`. The `..` would then move to the call-site, making exclusive-range explicit in the API

File: runtime/runtime/src/balance_checker.rs (hunk 1 above)
Comment (matklad) [reply runtime/runtime/src/balance_checker.rs:93]: Ah, we actually have that as `safe_add_balance_apply!` macro

**Review by matklad — APPROVED** [review]

**Comment by Longarithm** [comment]
This is awesome, I think it should help with resolving https://github.com/near/nearcore/issues/6308 as well.

**Comment by mina86** [comment]
Indeed.  I’ve run into this independently and have a couple more PRs which should push that Issue further still.

**Review by nikurt — COMMENTED** [review]
LGTM

**Pull request merged**